### PR TITLE
Ethflow orders in get_user_order()

### PR DESCRIPTION
**Pull request by josojo**
Since my previous [PR](https://github.com/cowprotocol/services/pull/583/files), the query get_user_order is no longer performant. Since the index `user_order_creation_timestamp` is a combination of user and creation_timestamp, my left outer join + additional where condition lowered the performance of the query significantly.

I see two ways to fix it:
- serve ethflow-orders and normal orders via different API requests and combine them in the front-end
- have non-optimized queries in the backend, hoping that they are still fast enough
(kudos to @vkgnosis who foresaw the challenge already 3 months ago during a design discussion. Unfortunately, I forgot about it)

This PR implements the second alternative:

I think if we have an index on the 'owner' instead the index on the combination of owner and creation_timestamp, then I guess our queries are still sufficiently fast, as the number of orders per users are not so huge and can be sorted by the creation_timestamp in reasonable time. Espeically for most non-bot users. We can combine this approach with limiting the amount of orders per user.
I don't know, but I really hope that its still fast enough. I used the test in the code and got the following results:
```
I created 1 users with  each 1000 orders and the test query within the code took `Duration { secs: 0, nanos: 6100000 }`
I created 64 users with  each 1000 orders and the test query within the code took  'Duration { secs: 0, nanos: 5461000 }'
I created 128 users with each 1000 orders and the test query within the code took `Duration { secs: 0, nanos: 4977000 }`
I created 240 users with  each 1000 orders and the test query within the code took `Duration { secs: 0, nanos: 5079000 }`
```
=>We can see that the performance is not really impacted by more users. Assuming, we will never have more than 1000 orders per user, the query should be still performant enough.

For reference, the indexing over owner+timstamp was introduced here: 
https://github.com/gnosis/gp-v2-services/pull/1020

For a complete overview, here is the combined log of the orders table:
```
-- Table: public.orders

-- DROP TABLE IF EXISTS public.orders;

CREATE TABLE IF NOT EXISTS public.orders
(
    uid bytea NOT NULL,
    owner bytea NOT NULL,
    creation_timestamp timestamp with time zone NOT NULL,
    sell_token bytea NOT NULL,
    buy_token bytea NOT NULL,
    sell_amount numeric(78,0) NOT NULL,
    buy_amount numeric(78,0) NOT NULL,
    valid_to bigint NOT NULL,
    fee_amount numeric(78,0) NOT NULL,
    kind orderkind NOT NULL,
    partially_fillable boolean NOT NULL,
    signature bytea NOT NULL,
    cancellation_timestamp timestamp with time zone,
    receiver bytea,
    app_data bytea NOT NULL,
    signing_scheme signingscheme NOT NULL,
    settlement_contract bytea NOT NULL,
    sell_token_balance selltokensource NOT NULL,
    buy_token_balance buytokendestination NOT NULL,
    full_fee_amount numeric(78,0) NOT NULL,
    is_liquidity_order boolean NOT NULL,
    CONSTRAINT orders_pkey PRIMARY KEY (uid)
)

TABLESPACE pg_default;

ALTER TABLE IF EXISTS public.orders
    OWNER to mainnet_admin;

REVOKE ALL ON TABLE public.orders FROM readonly;

GRANT SELECT ON TABLE public.orders TO gp_readonly;

GRANT ALL ON TABLE public.orders TO mainnet_admin;

GRANT SELECT ON TABLE public.orders TO readonly;
-- Index: order_creation_timestamp

-- DROP INDEX IF EXISTS public.order_creation_timestamp;

CREATE INDEX IF NOT EXISTS order_creation_timestamp
    ON public.orders USING btree
    (creation_timestamp ASC NULLS LAST)
    TABLESPACE pg_default;
-- Index: order_valid_to

-- DROP INDEX IF EXISTS public.order_valid_to;

CREATE INDEX IF NOT EXISTS order_valid_to
    ON public.orders USING btree
    (valid_to ASC NULLS LAST)
    TABLESPACE pg_default;
-- Index: user_order_creation_timestamp

-- DROP INDEX IF EXISTS public.user_order_creation_timestamp;

CREATE INDEX IF NOT EXISTS user_order_creation_timestamp
    ON public.orders USING btree
    (owner ASC NULLS LAST, creation_timestamp DESC NULLS FIRST)
    TABLESPACE pg_default;
-- Index: user_valid_to

-- DROP INDEX IF EXISTS public.user_valid_to;

CREATE INDEX IF NOT EXISTS user_valid_to
    ON public.orders USING btree
    (valid_to ASC NULLS LAST)
    TABLESPACE pg_default;
-- Index: version_idx

-- DROP INDEX IF EXISTS public.version_idx;

CREATE INDEX IF NOT EXISTS version_idx
    ON public.orders USING btree
    (settlement_contract ASC NULLS LAST)
    TABLESPACE pg_default;
```

### Test Plan

### CODE REVIEW

**Comment by josojo** [comment]
>We can likely solve this with a new index. 

I don't think so. One can not create an index above two tables (orders and onchain_orders). https://stackoverflow.com/questions/16402225/index-spanning-multiple-tables-in-postgresql

Regarding:  https://github.com/cowprotocol/services/issues/607, I think it's really different, as one can just create another index over receiver (+ creation_time) and one can easily keep the performance.

From a requirement perspective, I can confirm that we want the ethflow orders behind the normal user order endpoints

File: crates/database/src/orders.rs [hunk 1]
```diff
@@ -401,13 +401,18 @@ pub fn user_orders<'a>(
     // queries are taking too long in practice.
     #[rustfmt::skip]
     const QUERY: &str = const_format::concatcp!(
-"SELECT ", ORDERS_SELECT,
+"(SELECT ", ORDERS_SELECT,
+" FROM ", ORDERS_FROM,
+" LEFT OUTER JOIN onchain_placed_orders onchain_o on onchain_o.uid = o.uid",
+" WHERE o.owner = $1) ",
```
Comment (vkgnosis) [inline crates/database/src/orders.rs:407]: can't this be
```
 WHERE o.owner = $1 OR onchain_o.sender = $1
```
?

**Review by vkgnosis — APPROVED** [review]
I've thought about this for a while. I think this solution is good enough for me because we don't expect to have users with more orders than what we can reasonably hold in memory at the same time (which is needed in the background for this approach).

There are other ways to achieve a more correct solution but they are harder to design and I don't want to optimize this prematurely and delay this change.

File: crates/database/src/orders.rs (hunk 1 above)
Comment (josojo) [reply crates/database/src/orders.rs]: test result with UNION: 
test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 36 filtered out; finished in 310.63s

Test result without UNION:
test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 36 filtered out; finished in 328.50s

In the end, I went for your solution, as the test result difference is not big and might be due to other factors such as local usage.

Comment (vkgnosis) [reply crates/database/src/orders.rs:407]: I looked at it in the query analyzer.  For some reason it doesn't use the right index without UNION. So let's keep it.

File: database/sql/V034__create_index_order_owner.sql [hunk 2]
```diff
@@ -0,0 +1,4 @@
+-- replace index for effective order searching
+DROP INDEX public.user_order_creation_timestamp;
+
+CREATE INDEX order_owner ON public.orders USING HASH (owner);
```
Comment (nlordell) [inline database/sql/V034__create_index_order_owner.sql:4]:
```suggestion
DROP INDEX user_order_creation_timestamp;

CREATE INDEX order_owner ON orders USING HASH (owner);
```

Is there a reason we need `public` schema here? We don't use it in other migration scripts and would slightly prefer not to have it for consistency with our other SQL scripts.

File: crates/database/src/orders.rs [hunk 3]
```diff
@@ -781,14 +788,66 @@ mod tests {
         assert!(get_order(&mut db, 2).await.is_some());
     }
 
+    // This test is only for the PR. Can be removed before merging...
```
Comment (nlordell) [inline crates/database/src/orders.rs:791]: I would leave it in - its a manual test and may come in handy in the future.

**Review by nlordell — APPROVED** [review]

File: database/sql/V034__create_index_order_owner.sql (hunk 2 above)
Comment (vkgnosis) [reply database/sql/V034__create_index_order_owner.sql:4]: I didn't notice this while reviewing. Agree with the suggestion.

File: crates/database/src/orders.rs [hunk 4]
```diff
@@ -781,14 +788,66 @@ mod tests {
...
+                };
+                insert_order(&mut db, &order).await.unwrap();
+            }
+        }
+
+        let now = Utc::now();
+        let _result = user_orders(&mut db, &ByteArray([2u8; 20]), 10, Some(10)).await;
+        let time_diff = Utc::now() - now;
+        println!("{:?}", time_diff);
+        assert!(time_diff < chrono::Duration::seconds(1));
```
Comment (vkgnosis) [inline crates/database/src/orders.rs:840]:
```suggestion
        let now = std::time::Instant::now();
        let _result = user_orders(&mut db, &ByteArray([2u8; 20]), 10, Some(10)).await;
        let time_diff = now.elapsed();
        println!("{:?}", elapsed);
        assert!(elapsed < std::time::Duration::from_secs(1));
```
Prefer std types when they accomplish the same thing.

**Pull request merged**